components/encryption/src/master_key/kms.rs

```diff
@@ -52,28 +91,35 @@ impl AwsKms {
 
         Ok(AwsKms {
             client,
-            current_key_id: config.key_id.clone(),
+            current_key_id: KeyId::new(config.key_id)?,
             runtime,
-            _not_send: PhantomData::default(),
+            region: config.region,
+            endpoint: config.endpoint,
         })
     }
 
-    fn check_config(config: &KmsConfig) -> Result<()> {
-        if config.key_id.is_empty() {
-            return Err(box_err!("KMS key id can not be empty"));
-        }
-        Ok(())
+    pub fn new(config: KmsConfig) -> Result<AwsKms> {
+        // We must create our own dispatcher
+        // See https://github.com/tikv/tikv/issues/7236
+        let dispatcher = HttpClient::new().unwrap();
+        Self::new_with_dispatcher(config, dispatcher)
+    }
+}
+
+impl KmsProvider for AwsKms {
+    fn name(&self) -> &[u8] {
+        AWS_KMS_VENDOR_NAME
     }
 
     // On decrypt failure, the rule is to return WrongMasterKey error in case it is possible that
     // a wrong master key has been used, or other error otherwise.
-    fn decrypt(&mut self, ciphertext: &[u8]) -> Result<Vec<u8>> {
+    fn decrypt(&mut self, ciphertext: EncryptedRef) -> Result<Vec<u8>> {
```

Comment:
rename the type `EncryptedRef` to `CiphertextKey`, if you want to use the newtype ideom. Definition of the type should be moved to this file, as it is KMS specific.

Reply:
It would be nice to also define a new type for the plaintext key.

Reply:
well, in this context it is the encrypted data key. it is more clear if the type name is specific.